File: 2025-1/sesion-de-clase-24/java/src/main/java/CuentaBancaria.java


/**
 *
 * @author eric
 */
public class CuentaBancaria {
    private double saldo = 500.00;
    
    public synchronized void retirar(String nombreCliente, double monto) {
        while (this.saldo < monto) {
            System.out.println(nombreCliente + " quiere retirar " + monto + " pero no hay sificiente, esperando...");
            try {
                wait();
            }
            catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
            }
        }
               
        System.out.println("El cliente " + nombreCliente + " va a retirar " + monto);
//        try {
//            Thread.sleep(1000);
//        }
//        catch (InterruptedException ex) {
//            Thread.currentThread().interrupt();
//        }

        this.saldo -= monto;
        System.out.println(nombreCliente + " completo el retiro, queda " + this.saldo);
    }
    
    public synchronized void depositar(String nombreCliente, double monto) {
        System.out.println("El cliente " + nombreCliente + " deposito " + monto);
        
//        try {
//            Thread.sleep(5000);
//        }
//        catch (InterruptedException ex) {
//            Thread.currentThread().interrupt();
//        }
        
        saldo += monto;
        System.out.println("Saldo despues del deposito: " + this.saldo);
        notifyAll();
    }
}

File: 2026-1/sesion-de-clase-24/net/sesion-de-clase-24/Ejercicio4/CuentaBancaria.cs
using System;
using System.Threading;

namespace Ejercicio4;

public class CuentaBancaria
{
    private readonly object lockObject = new object();

    public double Saldo { get; set; } = 500.00;

    public void Retirar(string nombreCliente, double monto)
    {
        lock (lockObject)
        {
            while (Saldo < monto)
            {
                Console.WriteLine($"{nombreCliente} quiere retirar {monto} pero no hay sificiente, esperando...");
                Monitor.Wait(lockObject);
            }

            Console.WriteLine($"El cliente {nombreCliente} va a retirar {monto}");
            Saldo -= monto;
            Console.WriteLine($"{nombreCliente} completo el retiro, queda {Saldo}");
        }
    }

    public void Depositar(string nombreCliente, double monto)
    {
        lock (lockObject)
        {
            Console.WriteLine($"El cliente {nombreCliente} deposito {monto}");
            Saldo += monto;
            Console.WriteLine($"Saldo despues del deposito: {Saldo}");
            Monitor.PulseAll(lockObject);
        }
    }
}
